File: samples/Java/stable-marriage-problem.java
import java.util.*;

public class Stable {
    static List<String> guys = Arrays.asList(
            new String[]{
        "abe", "bob", "col", "dan", "ed", "fred", "gav", "hal", "ian", "jon"});
    static List<String> girls = Arrays.asList(
            new String[]{
        "abi", "bea", "cath", "dee", "eve", "fay", "gay", "hope", "ivy", "jan"});
    static Map<String, List<String>> guyPrefers =
            new HashMap<String, List<String>>(){{
        put("abe",
            Arrays.asList("abi", "eve", "cath", "ivy", "jan", "dee", "fay",
            "bea", "hope", "gay"));
        put("bob",
            Arrays.asList("cath", "hope", "abi", "dee", "eve", "fay", "bea",
            "jan", "ivy", "gay"));
        put("col",
            Arrays.asList("hope", "eve", "abi", "dee", "bea", "fay", "ivy",
            "gay", "cath", "jan"));
        put("dan",
            Arrays.asList("ivy", "fay", "dee", "gay", "hope", "eve", "jan",
            "bea", "cath", "abi"));
        put("ed",
            Arrays.asList("jan", "dee", "bea", "cath", "fay", "eve", "abi",
            "ivy", "hope", "gay"));
        put("fred",
            Arrays.asList("bea", "abi", "dee", "gay", "eve", "ivy", "cath",
            "jan", "hope", "fay"));
        put("gav",
            Arrays.asList("gay", "eve", "ivy", "bea", "cath", "abi", "dee",
            "hope", "jan", "fay"));
        put("hal",
            Arrays.asList("abi", "eve", "hope", "fay", "ivy", "cath", "jan",
            "bea", "gay", "dee"));
        put("ian",
            Arrays.asList("hope", "cath", "dee", "gay", "bea", "abi", "fay",
            "ivy", "jan", "eve"));
        put("jon",
            Arrays.asList("abi", "fay", "jan", "gay", "eve", "bea", "dee",
            "cath", "ivy", "hope"));
    }};
    static Map<String, List<String>> girlPrefers =
            new HashMap<String, List<String>>(){{
        put("abi",
            Arrays.asList("bob", "fred", "jon", "gav", "ian", "abe", "dan",
            "ed", "col", "hal"));
        put("bea",
            Arrays.asList("bob", "abe", "col", "fred", "gav", "dan", "ian",
            "ed", "jon", "hal"));
        put("cath",
            Arrays.asList("fred", "bob", "ed", "gav", "hal", "col", "ian",
            "abe", "dan", "jon"));
        put("dee",
            Arrays.asList("fred", "jon", "col", "abe", "ian", "hal", "gav",
            "dan", "bob", "ed"));
        put("eve",
            Arrays.asList("jon", "hal", "fred", "dan", "abe", "gav", "col",
            "ed", "ian", "bob"));
        put("fay",
            Arrays.asList("bob", "abe", "ed", "ian", "jon", "dan", "fred",
            "gav", "col", "hal"));
        put("gay",
            Arrays.asList("jon", "gav", "hal", "fred", "bob", "abe", "col",
            "ed", "dan", "ian"));
        put("hope",
            Arrays.asList("gav", "jon", "bob", "abe", "ian", "dan", "hal",
            "ed", "col", "fred"));
        put("ivy",
            Arrays.asList("ian", "col", "hal", "gav", "fred", "bob", "abe",
            "ed", "jon", "dan"));
        put("jan",
            Arrays.asList("ed", "hal", "gav", "abe", "bob", "jon", "col",
            "ian", "fred", "dan"));
    }};
    public static void main(String[] args){
        Map<String, String> matches = match(guys, guyPrefers, girlPrefers);
        for(Map.Entry<String, String> couple:matches.entrySet()){
            System.out.println(
                    couple.getKey() + " is engaged to " + couple.getValue());
        }
        if(checkMatches(guys, girls, matches, guyPrefers, girlPrefers)){
            System.out.println("Marriages are stable");
        }else{
            System.out.println("Marriages are unstable");
        }
        String tmp = matches.get(girls.get(0));
        matches.put(girls.get(0), matches.get(girls.get(1)));
        matches.put(girls.get(1), tmp);
        System.out.println(
                girls.get(0) +" and " + girls.get(1) + " have switched partners");
        if(checkMatches(guys, girls, matches, guyPrefers, girlPrefers)){
            System.out.println("Marriages are stable");
        }else{
            System.out.println("Marriages are unstable");
        }
    }

    private static Map<String, String> match(List<String> guys,
            Map<String, List<String>> guyPrefers,
            Map<String, List<String>> girlPrefers){
        Map<String, String> engagedTo = new TreeMap<String, String>();
        List<String> freeGuys = new LinkedList<String>();
        freeGuys.addAll(guys);
        while(!freeGuys.isEmpty()){
            String thisGuy = freeGuys.remove(0); //get a load of THIS guy
            List<String> thisGuyPrefers = guyPrefers.get(thisGuy);
            for(String girl:thisGuyPrefers){
                if(engagedTo.get(girl) == null){//girl is free
                    engagedTo.put(girl, thisGuy); //awww
                    break;
                }else{
                    String otherGuy = engagedTo.get(girl);
                    List<String> thisGirlPrefers = girlPrefers.get(girl);
                    if(thisGirlPrefers.indexOf(thisGuy) <
                            thisGirlPrefers.indexOf(otherGuy)){
                        //this girl prefers this guy to the guy she's engaged to
                        engagedTo.put(girl, thisGuy);
                        freeGuys.add(otherGuy);
                        break;
                    }//else no change...keep looking for this guy
                }
            }
        }
        return engagedTo;
    }

    private static boolean checkMatches(List<String> guys, List<String> girls,
            Map<String, String> matches, Map<String, List<String>> guyPrefers,
            Map<String, List<String>> girlPrefers) {
        if(!matches.keySet().containsAll(girls)){
            return false;
        }

        if(!matches.values().containsAll(guys)){
            return false;
        }

        Map<String, String> invertedMatches = new TreeMap<String, String>();
        for(Map.Entry<String, String> couple:matches.entrySet()){
            invertedMatches.put(couple.getValue(), couple.getKey());
        }

        for(Map.Entry<String, String> couple:matches.entrySet()){
            List<String> shePrefers = girlPrefers.get(couple.getKey());
            List<String> sheLikesBetter = new LinkedList<String>();
            sheLikesBetter.addAll(shePrefers.subList(0, shePrefers.indexOf(couple.getValue())));
            List<String> hePrefers = guyPrefers.get(couple.getValue());
            List<String> heLikesBetter = new LinkedList<String>();
            heLikesBetter.addAll(hePrefers.subList(0, hePrefers.indexOf(couple.getKey())));

            for(String guy : sheLikesBetter){
                String guysFinace = invertedMatches.get(guy);
                List<String> thisGuyPrefers = guyPrefers.get(guy);
                if(thisGuyPrefers.indexOf(guysFinace) >
                        thisGuyPrefers.indexOf(couple.getKey())){
                    System.out.printf("%s likes %s better than %s and %s"
                            + " likes %s better than their current partner\n",
                       couple.getKey(), guy, couple.getValue(),
                       guy, couple.getKey());
                    return false;
                }
            }

            for(String girl : heLikesBetter){
                String girlsFinace = matches.get(girl);
                List<String> thisGirlPrefers = girlPrefers.get(girl);
                if(thisGirlPrefers.indexOf(girlsFinace) >
                        thisGirlPrefers.indexOf(couple.getValue())){
                    System.out.printf("%s likes %s better than %s and %s"
                            + " likes %s better than their current partner\n",
                       couple.getValue(), girl, couple.getKey(),
                       girl, couple.getValue());
                    return false;
                }
            }
        }
        return true;
    }
}

File: samples/Haskell/stable-marriage-problem.hs
import Data.List
import Control.Monad
import Control.Arrow
import Data.Maybe

mp = map ((head &&& tail). splitNames)
   ["abe: abi, eve, cath, ivy, jan, dee, fay, bea, hope, gay",
    "bob: cath, hope, abi, dee, eve, fay, bea, jan, ivy, gay",
    "col: hope, eve, abi, dee, bea, fay, ivy, gay, cath, jan",
    "dan: ivy, fay, dee, gay, hope, eve, jan, bea, cath, abi",
    "ed: jan, dee, bea, cath, fay, eve, abi, ivy, hope, gay",
    "fred: bea, abi, dee, gay, eve, ivy, cath, jan, hope, fay",
    "gav: gay, eve, ivy, bea, cath, abi, dee, hope, jan, fay",
    "hal: abi, eve, hope, fay, ivy, cath, jan, bea, gay, dee",
    "ian: hope, cath, dee, gay, bea, abi, fay, ivy, jan, eve",
    "jon: abi, fay, jan, gay, eve, bea, dee, cath, ivy, hope"]

fp = map ((head &&& tail). splitNames)
    ["abi: bob, fred, jon, gav, ian, abe, dan, ed, col, hal",
     "bea: bob, abe, col, fred, gav, dan, ian, ed, jon, hal",
     "cath: fred, bob, ed, gav, hal, col, ian, abe, dan, jon",
     "dee: fred, jon, col, abe, ian, hal, gav, dan, bob, ed",
     "eve: jon, hal, fred, dan, abe, gav, col, ed, ian, bob",
     "fay: bob, abe, ed, ian, jon, dan, fred, gav, col, hal",
     "gay: jon, gav, hal, fred, bob, abe, col, ed, dan, ian",
     "hope: gav, jon, bob, abe, ian, dan, hal, ed, col, fred",
     "ivy: ian, col, hal, gav, fred, bob, abe, ed, jon, dan",
     "jan: ed, hal, gav, abe, bob, jon, col, ian, fred, dan"]

splitNames = map (takeWhile(`notElem`",:")). words

pref x y xs = fromJust (elemIndex x xs) < fromJust (elemIndex y xs)

task ms fs = do
 let
  jos = fst $ unzip ms

  runGS  es js ms = do
    let (m:js') = js
	(v:vm') = case lookup m ms of
		    Just xs -> xs
		    _ -> []
	vv = fromJust $ lookup v fs
	m2 = case lookup v es of
		Just e -> e
		_ -> ""
	ms1 = insert (m,vm') $ delete (m,v:vm') ms

    if null js then do
	putStrLn ""
	putStrLn "=== Couples ==="
	return es

      else if null m2 then
	do putStrLn $ v ++ " with " ++ m
	   runGS ( insert (v,m) es ) js' ms1
	
      else if pref m m2 vv then
	do putStrLn $ v ++ " dumped " ++ m2 ++ " for " ++ m
	   runGS ( insert (v,m) $ delete (v,m2) es ) (if not $ null vm' then js'++[m2] else js') ms1

      else runGS es (if not $ null js' then js'++[m] else js') ms1

 cs <- runGS [] jos ms

 mapM_ (\(f,m) ->  putStrLn $ f ++ " with " ++ m ) cs
 putStrLn ""
 checkStab cs

 putStrLn ""
 putStrLn "Introducing error: "
 let [r1@(a,b), r2@(p,q)] = take 2 cs
     r3 = (a,q)
     r4 = (p,b)
     errcs =  insert r4. insert r3. delete r2 $ delete r1 cs
 putStrLn $ "\tSwapping partners of " ++ a ++ " and " ++ p
 putStrLn $ (\((a,b),(p,q)) -> "\t" ++ a ++ " is now with " ++ b ++ " and " ++ p ++ " with " ++ q) (r3,r4)
 putStrLn ""
 checkStab errcs

checkStab es = do
  let
    fmt (a,b,c,d) =  a ++ " and " ++ b ++ " like each other better than their current partners " ++ c ++ " and " ++ d
    ies = uncurry(flip zip) $ unzip es  -- es = [(fem,m)]  & ies = [(m,fem)]
    slb = map (\(f,m)-> (f,m, map (id &&& fromJust. flip lookup ies). fst.break(==m). fromJust $ lookup f fp) ) es
    hlb = map (\(f,m)-> (m,f, map (id &&& fromJust. flip lookup es ). fst.break(==f). fromJust $ lookup m mp) ) es
    tslb = concatMap (filter snd. (\(f,m,ls) ->
				map (\(m2,f2) ->
				      ((f,m2,f2,m), pref f f2 $ fromJust $ lookup m2 mp)) ls)) slb
    thlb = concatMap (filter snd. (\(m,f,ls) ->
				map (\(f2,m2) ->
				      ((m,f2,m2,f), pref m m2 $ fromJust $ lookup f2 fp)) ls)) hlb
    res = tslb ++ thlb

  if not $ null res then do
    putStrLn "Marriages are unstable, e.g.:"
    putStrLn.fmt.fst $ head res

   else putStrLn "Marriages are stable"
